binr/r2r/r2r.c

```diff
@@ -41,23 +41,88 @@ static int help(bool verbose) {
 		" -r [radare2] path to radare2 executable (default is "RADARE2_CMD_DEFAULT")\n"
 		" -m [rasm2]   path to rasm2 executable (default is "RASM2_CMD_DEFAULT")\n"
 		" -f [file]    file to use for json tests (default is "JSON_TEST_FILE_DEFAULT")\n"
+		" -C [dir]     chdir before running r2r (default follows executable symlink + test/new\n"
 		"\n"
 		"OS/Arch for archos tests: "R2R_ARCH_OS"\n");
 	}
 	return 1;
 }
 
+static bool r2r_chdir(const char *argv0) {
+#if __UNIX__
+	if (r_file_is_directory ("db")) {
+		return true;
+	}
+	char src_path[PATH_MAX];
+	char *r2r_path = r_file_path (argv0);
+	bool found = false;
+	if (readlink (r2r_path, src_path, sizeof (src_path)) != -1) {
+		char *p = strstr (src_path, R_SYS_DIR "binr"R_SYS_DIR"r2r"R_SYS_DIR"r2r");
+		if (p) {
+			*p = 0;
+			strcat (src_path, R_SYS_DIR"test"R_SYS_DIR"new");
+			if (r_file_is_directory (src_path)) {
+				(void)chdir (src_path);
+				eprintf ("Running from %s\n", src_path);
+				found = true;
+			}
+		}
+	}
+	free (r2r_path);
+	return found;
+#else
+	return false;
+#endif
+}
+
+// TODO: move into libr/util/file.c
+static bool r_file_is_root(const char *cwd) {
+#if __WINDOWS__
+	return r_str_endswith (cwd, ":" R_SYS_DIR);
+#else
+	return !strcmp (cwd, R_SYS_DIR);
+#endif
+}
+
+static bool r2r_chdir_fromtest(const char *test_path) {
+	char *abs_test_path = r_file_abspath (test_path);
+	if (!r_file_is_directory (abs_test_path)) {
+		char *last_slash = (char *)r_str_lchr (abs_test_path, R_SYS_DIR[0]);
+		if (last_slash) {
+			*last_slash = 0;
+		}
+	}
+	(void)chdir (abs_test_path);
+	bool found = false;
+	char *cwd = NULL;
+	while (true) {
+		cwd = r_sys_getdir ();
+		if (r_file_is_root (cwd)) {
```

Comment:
I wouldn't check for root at all. If chdir fails for whatever reason you will still loop forever. Just do this instead:
```c
char *last_cwd = NULL;

while (true) {
    last_cwd = cwd;
    cwd = r_sys_getdir();
    if (last_cwd && !strcmp (last_cwd, cwd)) {
        free (last_cwd);
        free (cwd);
        break;
    }
    free (last_cwd);

....
```